botocore/credentials.py

```diff
@@ -835,161 +1016,295 @@ def __init__(self, load_config, client_creator, cache, profile_name,
+            refresher = fetcher.fetch_credentials
 
-        return cache_key.replace('/', '-')
+        expiration = _parse_if_needed(base_credentials['expiry_time'])
+        return RefreshableCredentials(
+            access_key=base_credentials['access_key'],
+            secret_key=base_credentials['secret_key'],
+            token=base_credentials['token'],
+            expiry_time=expiration,
+            method=self.METHOD,
+            refresh_using=refresher,
+            time_fetcher=_local_now
+        )
+
+    def _get_role_config(self, profile_name):
+        """Retrieves and validates the role configuration for the profile."""
+        profiles = self._loaded_config.get('profiles', {})
 
-    def _write_cached_credentials(self, creds, cache_key):
-        self.cache[cache_key] = creds
+        profile = profiles[profile_name]
+        source_profile = profile.get('source_profile')
+        role_arn = profile['role_arn']
+        credential_source = profile.get('credential_source')
+        mfa_serial = profile.get('mfa_serial')
+        external_id = profile.get('external_id')
+        role_session_name = profile.get('role_session_name')
 
-    def _get_role_config_values(self):
-        # This returns the role related configuration.
+        role_config = {
+            'role_arn': role_arn,
+            'external_id': external_id,
+            'mfa_serial': mfa_serial,
+            'role_session_name': role_session_name,
+            'source_profile': source_profile,
+            'credential_source': credential_source
+        }
+
+        # Either the credential source or the source profile must be
+        # specified, but not both.
+        if credential_source is not None and source_profile is not None:
+            raise InvalidConfigError(
+                error_msg=(
+                    'The profile "%s" contains both source_profile and '
+                    'credential_source.' % profile_name
+                )
+            )
+        elif credential_source is None and source_profile is None:
+            raise PartialCredentialsError(
+                provider=self.METHOD,
+                cred_var='source_profile or credential_source'
+            )
+        elif credential_source is not None:
+            self._validate_credential_source(
+                profile_name, credential_source)
+        else:
+            self._validate_source_profile(profile_name, source_profile)
+
+        return role_config
+
+    def _validate_credential_source(self, parent_profile, credential_source):
+        if self._credential_sourcer is None:
+            raise InvalidConfigError(error_msg=(
+                'The credential_source "%s" is specified in profile "%s", '
+                'but no source provider was configured.' % (
+                    credential_source, parent_profile)
+            ))
+        if not self._credential_sourcer.is_supported(credential_source):
+            raise InvalidConfigError(error_msg=(
```

Comment:
It would be helpful if we can also surface the list of canonical names with the `_credential_sourcer` knows what they are.  I ran into this my mistakenly using `EC2InstanceMetadata` instead of `Ec2InstanceMetadata`